modules/router/src/services/config.ts

```diff
@@ -81,6 +85,93 @@ export const getSwapFees = (
   });
 };
 
+export const getMappedAssets = (_assetId: string, _chainId: number): { assetId: string; chainId: number }[] => {
+  const routerAllowedSwaps = getConfig().allowedSwaps;
+
+  let pair: { assetId: string; chainId: number }[] = [{ assetId: _assetId, chainId: _chainId }];
+  routerAllowedSwaps.map((s) => {
+    if (s.fromAssetId.toLowerCase() === _assetId.toLowerCase() && s.fromChainId === _chainId) {
+      pair.push({ assetId: getAddress(s.toAssetId), chainId: s.toChainId });
+    } else if (s.toAssetId.toLowerCase() === _assetId.toLowerCase() && s.toChainId === _chainId) {
+      pair.push({ assetId: getAddress(s.fromAssetId), chainId: s.fromChainId });
+    }
+  });
+
+  const uniquePairs = Array.from(new Set(pair.map((s) => s.assetId))).map((x) => {
+    return pair.find((z) => z.assetId === x)!;
+  });
+  return uniquePairs;
+};
+
+export const onSwapGivenIn = async (
+  transferAmount: string,
+  fromAssetId: string,
+  fromChainId: number,
+  toAssetId: string,
+  toChainId: number,
+  routerSignerAddress: string,
+  chainReader: IVectorChainReader,
+): Promise<Result<{ priceImpact: string; amountOut: string }, ConfigServiceError>> => {
+  // get router balance for each chain for balances array to get the trade size.
+  // we will getOnChainBalance for routerSignerAddress
+  // get balance of token for fromChainId for router
+
+  const fromMappedAssets = getMappedAssets(fromAssetId, fromChainId);
+  const toMappedAssets = getMappedAssets(toAssetId, toChainId);
+
+  const mappedAssets = fromMappedAssets.concat(toMappedAssets);
+
+  const uniqueMappedAssets = Array.from(new Set(mappedAssets.map((s) => s.assetId))).map((x) => {
+    return mappedAssets.find((z) => z.assetId === x)!;
+  });
+
+  let balances = [];
+  for (var val of uniqueMappedAssets) {
+    console.log(val);
+    let onChainRouterBalance = await chainReader.getOnchainBalance(val.assetId, routerSignerAddress, val.chainId);
+    if (onChainRouterBalance.isError) {
+      return Result.fail(
+        new ConfigServiceError(ConfigServiceError.reasons.CouldNotGetAssetBalance, {
+          transferAmount,
+          fromAssetId,
+          fromChainId,
+          routerSignerAddress,
+        }),
+      );
+    }
+    balances.push(onChainRouterBalance);
+  }
+
+  const transferAmountBn = BigNumber.from(transferAmount);
+  const stableAmmAddress = getConfig().stableAmmAddress!;
+  const stableAmmProvider: JsonRpcProvider = new JsonRpcProvider(getConfig().stableAmmProvider!);
```

Comment:
Are these force unwraps "safe" here? As in, if the user does not define a valid env var for `STABLE_AMM_PROVIDER`, and calls this method `onSwapGivenIn`, is this going to error out in a way that's useful?

Or do we know for a fact they could not have gotten this far without those vars defined?

Reply:
Yeah, this is handled at the config layer itself.